### ISSUE TITLE
Download from arbitrary HTTP URLs with digest verification

### ISSUE THREAD

**Issue by luhring** [sigstore/sget#6]
### Context

Today `sget` can fetch content from OCI registries, but not from arbitrary HTTP URLs.

Importantly, `sget`'s OCI fetching never allows content to be downloaded without content verification. The OCI resource must either be signed and verified, or its digest must be specified by the user.

### Proposal

As a first step, let's ensure that HTTP downloads are supported **only if** the user provides an expected digest for the retrieved content. Until `sget` supports signature verification for HTTP retrievals, **providing a digest should be required**. Furthermore, `sget` should verify that the retrieved content's digest matches the user-provided digest before sharing the content with the user.

(This is [Phase 0 from the project Roadmap](https://github.com/sigstore/sget/blob/main/ROADMAP.md#phase-0).)

Proposed syntax:

```console
$ sget https://sh.rustup.rs --digest 'sha256:a3cb081f88a6789d104518b30d4aa410009cd08c3822a1226991d6cf0442a0f8'
```

### Open questions

1. Should `sget` allow HTTP downloads **without TLS**? And what about with invalid TLS? With a MITM attack, the content provided to the user would need to match the provided digest, even if the content was intercepted by a malicious actor, or else `sget` would prevent the user from receiving the content. However, there are other concerns left unaddressed.

**Comment by lukehinds** [comment]
So what comes to mind here, is how will users know the correct digest to use? 

Another factor is this labours a degree of maintenance on to the user. Right now they can just run `curl $site` now they need to check what is the correct digest for the very latest release and update their automation scripts.

I guess my point is, if you want folks to pivot away from curl , then adding on extra steps to manage is not likely going to convince them to change. Ideally you would want users to be able to run `alias curl="sget"` and have a seamless migration.

**Comment by luhring** [comment]
That's a good point, @lukehinds. This feature is intended to be one incremental step toward better security, but not representative of the eventual solution that we hope lures users away from `curl`.

>how will users know the correct digest to use?

Sometimes they won't know, in which case this feature doesn't help. But sometimes users would be able to simply copy an `sget ...` command from the software maintainer, such as on a GitHub `README.md`. We think this scenario will actually be somewhat common. In this situation, the user may feel comfortable trusting the README content, in which case they can ensure that the content they're retrieving hasn't been modified (maliciously or otherwise) since the time when the content existed on the maintainer's environment.

For cases where the user doesn't want to trust the command, this feature wouldn't be valuable.

So tl;dr: this is an incremental step, not a comprehensive solution.

**Comment by lukehinds** [comment]
> Sometimes they won't know, in which case this feature doesn't help. But sometimes users would be able to simply copy an sget ... command from the software maintainer, such as on a GitHub README.md. We think this scenario will actually be somewhat common. In this situation, the user may feel comfortable trusting the README content, in which case they can ensure that the content they're retrieving hasn't been modified (maliciously or otherwise) since the time when the content existed on the maintainer's environment.

That should really be avoided, as a compromise of the repository means someone swaps out a nefarious release with a nefarious digest [0] as both attack vectors are stored on the same medium. Also copy/paste operations can also very hazardous [1]

[0] https://micahflee.com/2016/02/backdoored-linux-mint-and-the-perils-of-checksums/
[1] https://thejh.net/misc/website-terminal-copy-paste

I would recommend you skip this phase as it has a low effort attack surface and can provide users with a false sense of security (which if it happens, damages the sigstore name).
